src/Core/Test.php

```diff
@@ -88,11 +98,88 @@ protected function runTearDown(TestResult $result, $action)
             foreach ($tearDowns as $tearDown) {
                 try {
                     $tearDown();
+                } catch (Throwable $e) {
+                    if ('passTest' === $action[0]) {
+                        $action = ['failTest', $this, $e];
+                    }
                 } catch (Exception $e) {
-                    $action = ['failTest', $this, $e];
+                    if ('passTest' === $action[0]) {
+                        $action = ['failTest', $this, $e];
+                    }
                 }
             }
         });
         call_user_func_array([$result, $action[0]], array_slice($action, 1));
     }
+
+    /**
+     * Set an error handler to handle errors within the test
+     *
+     * @param TestResult $result
+     * @param array      &$action
+     *
+     * @return array
+     */
+    protected function handleErrors(TestResult $result, &$action)
+    {
+        // first, unwind the current handler stack for later restoration
+        $handlers = [];
+
+        do {
```

Comment:
Wondering if we could extract this pattern of walking error handlers as well:

```php
function walkErrorHandlers(callable $cb) {
  do {
    if ($handler = set_error_handler(function() {})) {
      $cb($handler);
      restore_error_handler();
    }
    
    restore_error_handler();
  } while ($handler);
}
```

Thoughts?